File: 백준/Gold/1717. 집합의 표현/집합의 표현.cc
#include <bits/stdc++.h>
using namespace std;

int n, m;
vector<int> parent(1000002, -1);

int find(int x) {
    if (parent[x] < 0) return x;
    return parent[x] = find(parent[x]);
}

bool uni(int u, int v) {
    u = find(u);
    v = find(v);
    if (u == v) return false;

    if (u < v) parent[v] = u;
    else parent[u] = v;
    return true;
}

int main(void) {
    cin.tie(0)->sync_with_stdio(false);
    cin >> n >> m;
    while (m--) {
        int cmd, a, b;
        cin >> cmd >> a >> b;
        if (cmd == 0) {
            uni(a, b);
        } else {
            if (find(a) == find(b)) cout << "YES" << '\n';
            else cout << "NO" << '\n';
        }
    }
}

File: 백준/Gold/1717. 집합의 표현/집합의 표현.java
import java.io.*;
import java.util.*;

public class Main {

    static int n, m;
    static int[] parent = new int[1000002];

    static int find(int x) {
        if (parent[x] < 0) return x;
        return parent[x] = find(parent[x]);
    }

    static boolean uni(int u, int v) {
        u = find(u);
        v = find(v);
        if (u == v) return false;

        if (u < v) parent[v] = u;
        else parent[u] = v;
        return true;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        Arrays.fill(parent, -1);

        while (m-- > 0) {
            st = new StringTokenizer(br.readLine());
            int cmd = Integer.parseInt(st.nextToken());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());

            if (cmd == 0) {
                uni(a, b);
            } else {
                if (find(a) == find(b)) sb.append("YES\n");
                else sb.append("NO\n");
            }
        }

        System.out.print(sb);
    }
}
